smmregrid/gridtype.py

```diff
@@ -0,0 +1,214 @@
+import xarray as xr
+
+# default spatial dimensions and vertical coordinates
+DEFAULT_DIMS = {
+    'horizontal': ['i', 'j', 'x', 'y', 'lon', 'lat', 'longitude', 'latitude',
+                   'cell', 'cells', 'ncells', 'values', 'value', 'nod2', 'pix', 'elem',
+                   'nav_lon', 'nav_lat'],
+    'vertical': ['lev', 'nz1', 'nz', 'depth', 'depth_full', 'depth_half'],
+    'time': ['time']
+}
+
+
+class GridType:
+    """Fundamental GridType object"""
+
+    def __init__(self, dims, extra_dims=None, weights=None):
+        """
+        Initializes a GridType object carrying grid-specific information required by smmregrid.
+
+        Args:
+            dims (list): A list of default dimensions for the grid (e.g., ['time', 'lat', 'lon']).
+            extra_dims (dict, optional): A dictionary including keys 'vertical', 'time', and 'horizontal'
+                                          that can be used to extend the default dimensions. Defaults to None.
+            weights (any, optional): CDO weights used in regridding. It will initiate the object in a different
+                                         way assuming single-gridtype objects. Defaults to None.
+
+        Attributes:
+            dims (list): The dimensions defined for the grid.
+            horizontal_dims (list): The identified horizontal dimensions from the input.
+            vertical_dim (str or None): The identified vertical dimension, if applicable.
+            time_dims (list): The identified time dimensions from the input.
+            other_dims (list): The dimensions which are there but are not identified automatically.
+            variables (dict): A dictionary holding identified variables and their coordinates.
+            bounds (list): A list of bounds variables identified in the dataset.
+            masked (any): Placeholder for masked data (to be defined later).
+            weights (any): The weights associated with the grid, if provided.
+            cdo_weights (bool): A flag indicating if weights are provided.
+            weights_matrix (any): Placeholder for a weights matrix (to be defined later).
+            level_index (str): A string used to identify the index of the levels.
+
+        Raises:
+            ValueError: If multiple vertical dimensions are identified during initialization.
+        """
+
+        # key definitions
+        self.dims = dims
+        default_dims = self._handle_default_dimensions(extra_dims)
+        self.horizontal_dims = self._identify_dims('horizontal', default_dims)
+        self.vertical_dim = self._identify_dims('vertical', default_dims)
+        self.time_dims = self._identify_dims('time', default_dims)
+        self.other_dims = self._identify_other_dims()
+        self.variables = {}
+        self.bounds = []
+
+        # used by Regrid class
+        self.masked = None
+        self.weights = weights
+        self.cdo_weights = bool(weights)
```

Comment:
This is the same name used in GridInspector. (see comment above). Btw, why is this needed at all? One could always say `if self.weights` ... no?